mail-filter/opendkim: Improved OpenRC and systemd support

rseichter opened the pull request:
Michael Orlitzky and I have been working on improved OpenRC support for OpenDKIM. Our pull request has not yet been accepted upstream (not even acknowledged, in fact). The multi-file patch adds our changes to the existing Gentoo ebuild. Changes to the upstream build files include:

* `configure.ac` Added workaround to support @runstatedir@. On modern systems /run replaces /var/run, so using @runstatedir@ instead of @localstatedir@/run is preferable. Autoconf 2.70 will support @runstatedir@ , but it has not been released yet, so a workaround is required for version 2.61.

* `contrib/systemd` Use @runstatedir@ in opendkim.service.in, enable full @VARIABLE@ substitution, fix inconsistencies in the systemd service file, add tmpfiles.d configuration file.

* `opendkim/opendkim.conf.*.in` Use @runstatedir@ in PidFile paths, enable full @VARIABLE@ substitution.

orlitzky commented:
I think this line got dropped by accident:

```sh
sed -e 's:/var/db/dkim:/var/lib/opendkim:g' ...  opendkim/opendkim.conf.simple.in
```

But rather than put it back, I see later that you moved the conf additions into a separate file:

```
cat "${FILESDIR}/conf-additions" >>"${f}" || die
```

Instead of manually appending the contents to the sample config, why not use a patch? Then you can fix the `/var/db/dkim` -> `/var/lib/opendkim` path at the same time.

rseichter commented:
I decided to keep sed'ing the files that end up as documentation and install a ready-made `opendkim.conf file`. The upstream sample configuration is not ideal anyway, and for reasons we already discussed I would prefer to switch to a `local:` type socket default for new installations soonish.

orlitzky commented:
Nooooooooo we just spent a week moving our duplicated `$FILESDIR` files upstream. The `/var/db` -> `/var/lib` thing we can fix easily by adding `@LOCALSTATEDIR@` to the conf.simple.in file and list of variable substitutions, if you don't want to `sed` it.  The UMask line could easily go upstream right now, and the UserID entry has to eventually too (because right now, "opendkim" is hard-coded in the tmpfiles.d entry). After that the differences from upstream should be minimal -- probably just swapping out the inet socket for a local one.

rseichter commented:
> we just spent a week moving our duplicated `$FILESDIR` files upstream.

Providing a working `opendkim.conf` does not negate our work that went into the build, OpenRC and systemd files. How is removing the comments from a given sample file that does not even use the best settings and adding more settings of our own (be it via sed, cat, or patch) better than installing a complete config file?

All of our work could go upstream now. Alas, nobody there seems to care about the pull request so far. Once our changes are integrated, and once a new upstream release becomes available, we can polish the next ebuild version, but right now I have no idea if/when the OpenDKIM author will react.

orlitzky commented:
> Providing a working opendkim.conf does not negate our work that went into the build, OpenRC and systemd files. How is removing the comments from a given sample file that does not even use the best settings and adding more settings of our own (be it via sed, cat, or patch) better than installing a complete config file?

In the future, whenever someone wants to change the default config, it means that they'll be hacking our copy of the config with all of its hard-coded values rather than working on the upstream one, just like was done with the systemd and OpenRC service files. We didn't even know that there _were_ upstream systemd files, because someone before us copy/pasted a bad one into `$FILESDIR`.

There's really only three changes we need to make to the sample config:

  * Fix the `/var/db` paths
  * Set the UserID to opendkim
  * Set the UMask (in case someone wants to use a local socket)

The first two of those definitely need to go upstream, and the third is probably a safer upstream default than what they have now too. Patching them in (or using sed) makes it obvious that they're fixes applied on top of the upstream config, and allows other upstream changes to the sample config file to be merged seamlessly rather than having our copy fall out of date.  Patching also means we don't have to keep a new revision of `opendkim.conf` around every time an ebuild goes stable.

rseichter commented:
> In the future, whenever someone wants to change the default config, it means that they'll be hacking our copy of the config [...] rather than working on the upstream one

Given that I am the designated ebuild maintainer of OpenDKIM now (at least on paper), I wonder who but me is going to change the default config? 😉 Also, please consider that the last upstream release was made in four years (!) ago. This is not the Linux Kernel. OpenDKIM releases are infrequent and the default configuration is not prone to change often.

> There's really only three changes we need to make to the sample config

Sure, and that's happening, but there are also changes I want to make on top of that: Changing `Canonicalization` (which I already did) and `Socket`. The former may be accepted upstream, the latter might not be.

> Patching also means we don't have to keep a new revision of `opendkim.conf` around every time an ebuild goes stable.

Why would the contents of `opendkim.conf` need to change when an ebuild is stabilized? By the way, how is that stabilization thingy supposed to happen anyway?

klondi commented:
"I wonder who but me is going to change the default config?"
QA may, for example, do that.

"OpenDKIM releases are infrequent and the default configuration is not prone to change often." 
They released a beta on November 2018: https://github.com/trusteddomainproject/OpenDKIM/releases/tag/rel-opendkim-2-11-0-Beta2

"Why would the contents of opendkim.conf need to change when an ebuild is stabilized?"
Because paths, for example, may change in the distribution. It's uncommon but has happened in the past. A new revision may not be needed in such cases because the changes will only affect new installs.

"By the way, how is that stabilization thingy supposed to happen anyway?"
After some time in the main tree (usually a month) and supposing other conditions are met, you can request stabilization. See https://wiki.gentoo.org/wiki/Stable_request

Also if your openrc file will include grep, make sure to add sys-apps/grep to your ebuild RDEPENDS.

rseichter commented:
> QA may, for example, do that.

May, perhaps. Likely, not. I know what a good opendkim.conf file needs, and I don't expect anybody to change things without contacting me.

> They released a beta on November 2018

Yes, I am well aware. The last *stable* release however was four years ago, and at this time I don't plan on supporting beta releases in ebuilds.

I know that you are only pointing out what may happen, however unlikely, but since I am the one that would need to make changes, it does not worry me.

> After some time in the main tree (usually a month) and supposing other
> conditions are met, you can request stabilization.

Thanks, I'll look into it.

> Also if your openrc file will include grep, make sure to add sys-apps/grep to your ebuild RDEPENDS.

Only the ebuild uses grep, the OpenRC files don't.

klondi commented:
"Only the ebuild uses grep, the OpenRC files don't."
Then it should be defined on DEPEND.

orlitzky commented:
> >  "Only the ebuild uses grep, the OpenRC files don't."

>Then it should be defined on DEPEND.

It is used in the OpenRC service file, but the `./configure` script looks for it, so actually it should be in both.

My point about stabilization is that, once a stable ebuild uses the new `opendkim.conf` file, you aren't allowed to change that conf file. So, if you want to change it, you won't be able to without creating a separate revision of the conf file and a separate revision of the ebuild to use it. And if you decide to delete it... well, you can't, if any stable ebuilds are using it. So you have to wait at least a month to stabilize some OTHER version, and then remove all of the older ones, finally freeing up the (then-unused) conf file for deletion. It just creates more and more "paperwork."

And once you've moved it out of the upstream tree, people are going to blame you for everything in it. "Why did you change the default canonicalization? You broke my mom's server!" etc. And they have a point, because we should only be altering the upstream package when absolutely necessary. When they don't like the upstream config, you can at least tell them to complain there.  Basically, everything in `$FILESDIR` is a liability to your free time.

I realize that upstream development has stopped in this case, so maybe you're lucky and no one ever changes it again. But in general forking the configuration file is not something you'd do, and there's no especially good reason to do it here, so best to avoid the bad habit.

rseichter commented:
@orlitzky I think the want to achieve the same goal, but are willing to accept different grades of perfection to get there.

As for not forking the config file, I don't buy into that. I'd think long and hard about changing something as thoroughly refined as Postfix's default configuration, but I have no qualms correcting shortcomings in OpenDKIM's sample config. The project looks, as much as I hope I am wrong, pretty much abandoned upstream, beta release or not, so I say we change what we think we should.

Now, if you insist on making further changes to the PR, I'll not be grumpy. Most of the work is yours anyway, as we both know. Do what you feel you must do, if you really, really cannot live with my approach. No hard feelings. 🤓

orlitzky commented:
I'll still merge it, I just feel like it's a step backwards after I spent hours making sure that we could substitute things properly into the sample config. (And you ultimately don't need me -- the proxy-maint team would merge this eventually. I've CC'ed them on all the bugs.)

In any case, we need systemd users to test it first. Can you put back the local socket line from the sample config? The one that tells the user to place the socket under `@RUNSTATEDIR@/opendkim`? You will have to hard-code that value to `/run/opendkim` again I guess. Then, please ask someone on the systemd bug to test the new service/tmpfiles with a local socket. Everything should work out-of-the-box if they follow the directions in the sample config.

rseichter commented:
I'll try to look into this tomorrow. I't almost 02:00 local time, and I need to get some sleep.

juippis commented on mail-filter/opendkim/opendkim-2.10.3-r11.ebuild:
```diff
@@ -0,0 +1,205 @@
+
+src_install() {
+	default
+	prune_libtool_files
```
You're not inheriting this eclass, switch to suggested way: https://devmanual.gentoo.org/eclass-reference/ltprune.eclass/index.html

rseichter replied:
Will do.

rseichter commented:
> you ultimately don't need me -- the proxy-maint team would merge this eventually. I've CC'ed them on all the bugs.

They have a lot on their plate, and you know better than the proxy-maint team what this PR contains, given that you did most of the work yourself.

> Can you put back the local socket line from the sample config?

Ha, that one actually disappeared because of the pre-existing `sed` step that removes all comments. 🙄 I guess I might need to pull all our changes into src_install(). At least now I've had some sleep.

rseichter commented:
I have pulled all changes to `opendkim.conf` into the ebuild, and added the Socket comments manually. I don't know how else to do it if your heart is set on using the upstream config file and stripping all comments from it. 😉 At least the ebuild is now self-contained in that regard and there is no need to have config file revisions.

I've also added dependencies to `grep` and `sed`.

rseichter commented:
As we discussed, comments are no longer stripped from the upstream config file. The elog lines are cleaned up and point to the Wiki. Now to find someone who is willing to help by testing with systemd.

orlitzky commented:
I can't sensibly explain this because our policy about the `@system` set is so inconsistent, but you don't need to have `sys-apps/sed` in `DEPEND` when you only use it in the ebuild. On the other hand, grep _should_ be there (even though it's in `@system`) because the ./configure script looks for it and the OpenRC service script uses it.

The other stuff looks good. This line however doesn't do quite what you want it to:

```sh
	grep -iv '^(TrustAnchorFile|UMask|UserID)' \
		"${S}/opendkim/opendkim.conf.simple" >"${cf}" || die
```

that will kill the install if e.g. the upstream file doesn't contain anything *but* those three directives. The path to hell is paved with good intentions =)

Since the upstream file doesn't contain any of those directives at the moment, it's probably safe to leave out the grep entirely. I haven't tested, but even if the daemon refuses to start when a directive is specified twice, that would be a good hint to us to update the ebuild.

```
	sed -E -i -e 's:^(Canonicalization).+:\1\trelaxed/relaxed:' "${cf}" || die
```

Be sure to match the space after the parameter name here, in case there's ever a parameter named something like CanonicalizationType. For example

```
sed -e 's:^Canonicalization[[:space:]][[:space:]]*relaxed/simple$:Canonicalization\trelaxed/relaxed:' "${cf}"
```

I would also add a comment pointing to your upstream PR, in case some useful discussion happens there (or it gets merged, making the "sed" call obsolete).

orlitzky commented:
^ I forgot the `-i` in that sed line, but you get the idea.

rseichter commented:
> ```shell
> grep -iv '^(TrustAnchorFile|UMask|UserID)' \
> "${S}/opendkim/opendkim.conf.simple" >"${cf}" || die
> ```
>
> that will kill the install if e.g. the upstream file doesn't contain anything _but_ those three directives.

Are you quite sure? My grep-fu tells me that `grep -v pattern in >out` would, if all lines of `in` were matched by `pattern`, result in `out` being zero bytes, which is fine. I can't test this right now, but I am feeling confident. 😉

orlitzky commented:
The `-v` merely inverts the match, so grep will "fail" if there's nothing in the file that doesn't match your pattern. I tested to be sure:

```sh
$ cat foo.txt 
hello
hello
hello
$ grep -v hello foo.txt
$ echo $?
1
```

rseichter commented:
> The `-v` merely inverts the match, so grep will "fail" if there's nothing in the file that doesn't match your pattern.

I know what `-v`does, which is why I used it. 😜 What I did not expect is `grep` to consider "no matches" as an error in this case, but your test is clear on that. I'll change it.

rseichter commented:
Well, I have now wasted several hours trying to setup a systemd-based Gentoo in VirtualBox. emerge croaks on several packages, including glibc, and I won't keep trying anymore. My call for assistance on the proxy-maint mailing list has not been answered yet, but as it is Easter, people are probably occupied with other affairs.

orlitzky commented:
One last thing that I noticed on a fresh install (sorry): we need to call `tmpfiles_process()` from `tmpfiles.eclass` in `pkg_postinst()`. Otherwise, the directory `/run/opendkim` isn't created until you reboot, and thus new installs can't start the daemon with a local socket.

rseichter commented:
Does anybody else hear Limahl faintly singing "Never ending stoooryyy..."?

rseichter commented:
I have asked for help with systemd related testing via the linked bug and the proxy-maint mailing list, but that did not work out. I have also spent more time trying to get a systemd-based virtual Gentoo server running, and that did not work out either. What can be done to move this pull request along?

orlitzky commented:
I will eventually just merge it if we don't hear anything on the systemd front.

orlitzky commented:
Thanks for all your work on this Ralph, and thanks @klondi  for the OpenSSL patch. I merged it, and updated https://wiki.gentoo.org/wiki/OpenDKIM one last time to reflect the changes. If systemd is broken, I'm sure someone will file a bug =P
